server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -320,34 +333,46 @@ public void run() {
             }
             boolean nextRunScheduled = false;
             Runnable scheduleNextRun = new RunOnce(this::scheduleNextRunIfNecessary);
+            // Before we do anything, we're first going to make sure there is no in-flight request at this moment.
+            // If that's the case, we'll acquire the "lock", which prevents any other thread/instance from sending any requests.
+            if (inFlightRequest.compareAndSet(false, true) == false) {
+                logger.debug("Not allowed to send health info update request due to in-flight request, will try again.");
+                scheduleNextRun.run();
+                return;
+            }
             try {
                 List<HealthTracker.HealthProgress<?>> healthProgresses = getHealthProgresses();
                 if (healthProgresses.isEmpty()) {
                     // Next run will still be scheduled in the `finally` block.
+                    inFlightRequest.set(false);
                     return;
                 }
                 // Create builder and add the current value of each (changed) health tracker to the request.
                 var builder = new UpdateHealthInfoCacheAction.Request.Builder().nodeId(clusterService.localNode().getId());
                 healthProgresses.forEach(changedHealthInfo -> changedHealthInfo.updateRequestBuilder(builder));
 
-                var healthNodeId = lastSeenHealthNode.get();
                 var listener = ActionListener.<AcknowledgedResponse>wrap(response -> {
-                    // Don't update the latest health info if the health node has changed while this request was being processed.
-                    if (Objects.equals(healthNodeId, lastSeenHealthNode.get()) == false) {
-                        return;
+                    // Only record health progress if this monitoring instance hasn't been cancelled in the meantime.
+                    // This avoids any unwanted writes to the HealthTrackers' states after a new monitoring instance has possibly
+                    // already started.
+                    if (cancelled == false) {
+                        healthProgresses.forEach(HealthTracker.HealthProgress::recordProgressIfRelevant);
                     }
-                    healthProgresses.forEach(HealthTracker.HealthProgress::recordProgressIfRelevant);
+                    inFlightRequest.set(false);
```

Comment:
Can we combine this along with the `scheduleNextRun.run()`? It looks like they should be triggered at the same time, when the current run completes. Right?

Reply:
And rename it accordingly, of course.

Reply:
Hmm I don't think that will work because of this part:

https://github.com/elastic/elasticsearch/blob/6f34bb10660bb52774e7c05493f2bb3c362fd642/server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java#L338-L342

In that case, we don't want the `scheduleNextRun` to set the `inFlightRequest` to false, because we actually know for sure another thread has just sent a request. I do like the idea of moving the `inFlightRequest.set(false)` to one place so it's a bit easier to read/understand, but I can't really come up with a way that would work well. I could have two different `Runnable`s, the `scheduleNextRun` and one that wraps that one with the `.set(false)` statement, but I don't think that would make things a lot more readable tbh. What do you think?

Reply:
Hm, I see your point. We could consider keeping this as an exception maybe. Since it's out of the try, and use a runnable that will contain both statements on the rest of the code. Would that benefit us at all?

Reply:
I think from a stylistic POV that would indeed be better. But there's one more execution path that I'm not entirely sure about. Would it be possible for the `client.execute(...)` invocation to throw an exception _after_ it has actually sent the request (and will execute the listener after the request completes). In other words, would it be possible that we end up in the `catch` block but that the `listener` will be executed afterward?

Your proposed refactoring doesn't impact this, I think, but I only realized this after implementing your proposal.

Reply:
Let's dive into this. We will investigate 3 scenarios:
- `client.execute` succeeds in making a request
- `client.execute` is called but throws an exception (if possible)
- There is an exception before `client.execute` is called. 

Before we start, the catch block is not as relevant here, right? The `scheduleNextRun.run()` is being executed in the `finally` block. However, when an exception is thrown is important because it affects the value of the `nextRunScheduled` boolean flag which determines if the finally block will schedule the next run or not.

**Scenario 1**
If the `client.execute` succeeds then the `nextRunScheduled = true` will also succeed so the `scheduleNextRun.run()` will be executed only by the listener.
```
try {
    .......
    client.execute(UpdateHealthInfoCacheAction.INSTANCE, builder.build(), ActionListener.runAfter(listener, scheduleNextRun));
     nextRunScheduled = true;
} catch (Exception e) {
     logger.warn(() -> format("Failed to run scheduled health monitoring on thread pool [%s]", executor), e);
} finally {
      // If the next run isn't scheduled because for example the health info hasn't changed, we schedule it here.
      if (nextRunScheduled == false) {
            scheduleNextRun.run();
       }
}
```

Scenario 2
If the `client.execute` fails then the `nextRunScheduled` will be `false` meaning the `finally` block will be the one triggering the `scheduleNextRun.run()`. So, here, we will investigate if it's possible that listener will also be called. As you see there is a try-catch here that will delegate any failure to the listener, in this case the `client.execute` will not itself fail. Theoretically, we are safe. 
```
       try {
            doExecute(action, request, listener);
        } catch (Exception e) {
            assert false : new AssertionError(e);
            listener.onFailure(e);
        }
```

Scenario 3
If we fail anywhere before `client.execute` is called then the `nextRunScheduled` will be false and the finally block will save the day.

If in any case there is double invocation, we have wrapped the runnable with `RunOnce` which will protect us from a double invocation. 

Is this clear?

Reply:
Ahh right I forgot about the `RunOnce` part 😅. But besides that, your theory definitely makes sense, so thanks a lot for the elaborate explanation!

Reply:
Happy to! Keep asking these questions!!! It gives us the opportunity to see if we missed something!